### PR TITLE
feat: detect git conflicts on multi-user values changes

### LINKED ISSUE #867: [API/Console] Removed configuration keeps getting re-added to values repo
With more and more concurrent users using Otomi at the client, we are encountering the issue where configuration that has been deleted (in this case the Vault unseal config), keeps being re-introduced to the values repo while deploying other changes. This issue persists even after API and Console pods are killed.

For instance:

1. Vault unseal config is deleted in values
2. Drone pipeline runs
3. API and console pods are killed
4. User makes a change in Console (e.g. adding a secret)
5. Clicks deploy changes
6. Deleted Vault unseal config is re-introduced to values
7. Vault service is interrupted as we're not yet ready to deploy this configuration, config must be removed once again and Vault pod killed

I have asked the DevOps team to clear their browser cache, but observed the same thing happen to a user who opened Otomi in a private window, so we're stumped. It's still possible that the browser cache was at fault as I can't be sure if he opened the window before or after the API and Console pods were killed. Regardless, with more and more concurrent users it will be impossible to ensure everyone has cleared their browser cache in sync, and that a contaminated state won't make its way back to the API.
Below is a screenshot of a 3 user collision, with relevant commits highlighted:
Screenshot 2022-08-11 at 09 00 47

### PR DESCRIPTION
**Pull request by Morriz**
Forces api to allow one edit session at a time, started when a user makes any successful POST/PUT. Uses socket pubsub to realtime update users with notifications.

fixes #867 
fixes redkubes/unassigned-issues#446

## Checklist

- [ ] Architecture Design Records have been added as `adr/*.md` and appended to list in `adr/_index.md`, if applicable.
- [ ] The `values-schema.yaml` file and `test/**` fixtures have been updated to reflect code changes, if applicable.
- [ ] The OpenApi Schema from redkubes/otomi-api project is compatible with definitions from `values-schema.yaml` file, if applicable.
- [ ] Helm releases are meeting otomi's baseline security policies, if applicable.
- [ ] Helm chart and helmfile changes are tested against upgrade scenario, if applicable.

### CODE REVIEW

File: charts/team-ns/templates/_ingress.tpl [hunk 1]
```diff
@@ -114,6 +114,22 @@ metadata:
     nginx.ingress.kubernetes.io/auth-url: "http://oauth2-proxy.istio-system.svc.cluster.local/oauth2/auth"
     nginx.ingress.kubernetes.io/auth-signin: "https://auth.{{ $v.cluster.domainSuffix }}/oauth2/start?rd=/oauth2/redirect/$http_host$escaped_request_uri"
       {{- end }}
+    # websockets can stay open one hour
+    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
+    nginx.ingress.kubernetes.io/proxy-send-timeout: "3600"
+    # websocket upgrade snippet
+    nginx.ingress.kubernetes.io/server-snippets: |
+      location ~* (/ws/|socket.io) {
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_ingress.tpl:122]: a location paths are implementation specific. We can leave it like that for now, but it will have to be configurable once Ingress resource is configurable per service.

**Review by j-zimnowoda — APPROVED** [review]

**Comment by staticvoid255** [comment]
Really don't like this solution. See no reason why user should push the whole state to the API rather than just their changes plus their opinion of what the current state is as per their local cache - which the API can then reconcile against the values repo. In the case of a conflict the user could be prompted to drop changes or overwrite.

**Comment by Morriz** [comment]
> Not a fan of this solution. See no reason why user should push the whole state to the API rather than just their changes plus their opinion of what the current state is as per their local cache - which the API can then reconcile against the values repo. In the case of a conflict the user could be prompted to drop changes or overwrite.

We have a problem now, so we built something to alleviate it. We will build out our multi user scenario in a big epic, as that is not easy. Don't just think that git makes this easy.

**Comment by Morriz** [comment]
> Not a fan of this solution. See no reason why user should push the whole state to the API rather than just their changes plus their opinion of what the current state is as per their local cache - which the API can then reconcile against the values repo. In the case of a conflict the user could be prompted to drop changes or overwrite.

Reading this I see you don't have clarity on how it all works together. Clients have no "local cache". The api has one source of truth: the db, which gets deflated to a commit. In the current situation all users get to modify the db, and any second user will "taint" the intent of the first users modification. So we have to make sure one user only commits their changes, and nothing else, atomically. I have a design for multi-user approach with one in-mem db per user session, that could solve the tainting problem. But the bigger issue is what I told Jeho as well: we can't allow conflicts to arise as we can't support resolution of them (yet). So for now we have a hard requirement to not allow conflicts. In the future we can build  tooling to resolve conflicts in the console, so multi-user is possible again.

**Comment by staticvoid255** [comment]
Please talk with @j-zimnowoda as we have discussed multiple alternate solutions that are less detrimental to the UX. This solution is adding complexity instead of removing, and will only make it more difficult to address the root cause in the future. One of the solutions we came up with is to ditch the deploy changes button altogether, and to have changes deployed immediately on pressing the submit button of a form - the API will publish a revision number which the user will cache locally, and if that revision number changes (another user changes something) before the 2nd user has submitted their changes, then it will trigger a refresh on the 2nd users session. This approach limits disruption to other users sessions (losing changes) to only a single scope, rather than possibly losing many changes), and is a natural blocker to conflicts occurring in the first place.

I have seen this single-editor only approach in action and can guarantee it will result in significant disruption and annoyance among tenants as they will constantly step on each others toes.

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
I tried to deploy from this branch but I got an error
```

node:internal/modules/cjs/loader:936
  throw err;
  ^
Error: Cannot find module 'yaml'
Require stack:
- /home/app/stack/dist/src/common/utils.js
- /home/app/stack/dist/src/common/crypt.js
- /home/app/stack/dist/src/common/bootstrap.js
- /home/app/stack/dist/src/cmd/apply.js
- /home/app/stack/dist/src/cmd/index.js
- /home/app/stack/dist/src/otomi.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/home/app/stack/dist/src/common/utils.js:16:16)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/app/stack/dist/src/common/utils.js',
    '/home/app/stack/dist/src/common/crypt.js',
    '/home/app/stack/dist/src/common/bootstrap.js',
    '/home/app/stack/dist/src/cmd/apply.js',
    '/home/app/stack/dist/src/cmd/index.js',
    '/home/app/stack/dist/src/otomi.js'
  ]
}
```
See more about the way I deployed at: https://github.com/redkubes/otomi-core/actions/runs/3411754347/jobs/5676364106

Could you take a look?

File: charts/team-ns/templates/_ingress.tpl [hunk 2]
```diff
@@ -114,6 +115,22 @@ metadata:
     nginx.ingress.kubernetes.io/auth-url: "http://oauth2-proxy.istio-system.svc.cluster.local/oauth2/auth"
     nginx.ingress.kubernetes.io/auth-signin: "https://auth.{{ $v.cluster.domainSuffix }}/oauth2/start?rd=/oauth2/redirect/$http_host$escaped_request_uri"
       {{- end }}
+    # websockets can stay open one hour
+    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_ingress.tpl:119]: Is it possible to set those timeout only for otomi-api?

File: charts/team-ns/templates/_ingress.tpl [hunk 3]
```diff
@@ -114,6 +115,22 @@ metadata:
...
+    nginx.ingress.kubernetes.io/proxy-send-timeout: "3600"
+    # websocket upgrade snippet
+    nginx.ingress.kubernetes.io/server-snippets: |
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_ingress.tpl:122]: I looked into nginx-ingress controller documentation and it seems that they support websockets out of the box.
https://kubernetes.github.io/ingress-nginx/user-guide/miscellaneous/#websockets
Are you sure that this code snippet is necessary?

File: charts/team-ns/templates/_ingress.tpl [hunk 4]
```diff
@@ -114,6 +115,22 @@ metadata:
     nginx.ingress.kubernetes.io/auth-url: "http://oauth2-proxy.istio-system.svc.cluster.local/oauth2/auth"
     nginx.ingress.kubernetes.io/auth-signin: "https://auth.{{ $v.cluster.domainSuffix }}/oauth2/start?rd=/oauth2/redirect/$http_host$escaped_request_uri"
       {{- end }}
+    # websockets can stay open one hour
+    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
+    nginx.ingress.kubernetes.io/proxy-send-timeout: "3600"
+    # websocket upgrade snippet
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_ingress.tpl:121]: > If the NGINX ingress controller is exposed with a service type=LoadBalancer make sure the protocol between the loadbalancer and NGINX is TCP.

Please make a note in  https://otomi.io/docs/installation/prerequisites

File: src/common/k8s.ts [hunk 5]
```diff
@@ -59,13 +59,13 @@ export interface DeploymentState {
 }
 
 export const getDeploymentState = async (): Promise<DeploymentState> => {
-  if (env.DISABLE_SYNC) return {}
+  if (env.isDev && env.DISABLE_SYNC) return {}
```
Comment (j-zimnowoda) [inline src/common/k8s.ts:62]: Everywhere in code I see a pair `env.isDev && env.DISABLE_SYNC`

I think DISABLE_SYNC should always set to false if `env.isDev=false`, then you can use a very simple if statement through the code.
Also if DISABLE_SYNC is meant to use only in dev mode them perhaps it should have DEV_ prefix, i.e.: DEV_DISABLE_SYNC

File: src/common/utils.ts [hunk 6]
```diff
@@ -23,7 +24,7 @@ export const asArray = (args: string | string[]): string[] => {
 
 export const removeBlankAttributes = (obj: Record<string, any>): Record<string, any> => {
   const options: CleanOptions = {
-    emptyArrays: false,
+    emptyArrays: true,
```
Comment (j-zimnowoda) [inline src/common/utils.ts:27]: Please explain why?

File: src/common/values.ts [hunk 7]
```diff
@@ -102,29 +103,29 @@ const writeValuesToFile = async (
...
+    const notExists = !(await pathExists(targetPath))
     if (isSecretsFile && notExists) {
-      await writeFile(targetPath, objectToYaml(newValues))
+      await writeFile(targetPath, objectToYaml(values))
       await encrypt(targetPath)
       await decrypt(targetPath)
       return
     }
-    await writeFile(targetPath + suffix, objectToYaml(newValues))
+    await writeFile(targetPath + suffix, objectToYaml(values))
```
Comment (j-zimnowoda) [inline src/common/values.ts:123]: before values with removed blank attributes were saved but now you decided to save values with blank attributes. Why?

File: src/common/values.ts [hunk 8]
```diff
@@ -137,23 +138,27 @@ const writeValuesToFile = async (
...
+export const writeValues = async (inValues: Record<string, any>, overwrite = false): Promise<void> => {
   const d = terminal('common:values:writeValues')
-  d.debug('Writing values: ', values)
-  hasSops = existsSync(`${env.ENV_DIR}/.sops.yaml`)
-  // We relocated teamConfig.teams to teamConfig, so one time we might have to fall back to our previous location
-  // TODO:[teamConfig] deprecate this fallback sometime in the future
-  const teams = Object.keys((values.teamConfig?.teams ?? values.teamConfig) || {}).filter((i) => i !== 'teams')
+  d.debug('Writing values: ', inValues)
+  hasSops = await pathExists(`${env.ENV_DIR}/.sops.yaml`)
+  const teamConfig = (await loadYaml(`${env.ENV_DIR}/env/teams.yaml`)) as Record<string, Record<string, any>>
```
Comment (j-zimnowoda) [inline src/common/values.ts:145]: please add a comment in code that explains why teamConfig needs to be read from file while saving the inValues

File: .env.sample [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# will bypass docker but expects all binaries present on host
+IN_DOCKER=1
+VERBOSITY=2
+# will disable contacting the cluster as found in kube context:
```
Comment (j-zimnowoda) [inline .env.sample:4]: please add description env at `src/common/envalid.ts`

File: .values/env/teams/.gitinclude [hunk 10]
```diff
@@ -0,0 +1 @@
+*
```
Comment (j-zimnowoda) [inline .values/env/teams/.gitinclude:1]: why do we need this rule?

Comment (Morriz) [reply .values/env/teams/.gitinclude:1]: Placeholder for teams that are empty at start.

File: charts/team-ns/templates/argocd/argocd-project.yaml [hunk 11]
```diff
@@ -55,14 +55,16 @@ spec:
 {{- end }}
   - name: team-member
     description: Team member privileges to team-{{ $v.teamId }}
+{{- if or (has "argocd" $v.selfService.apps) $v.otomi.isMultitenant }}
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/argocd/argocd-project.yaml:58]: this is out of scope of this feature please make separate PR

File: schemas/gen-missing-crd-schemas.sh [hunk 12]
```diff
@@ -2,7 +2,8 @@
 
 # run from root
 
-cd schemas/
```
Comment (j-zimnowoda) [inline schemas/gen-missing-crd-schemas.sh:5]: Please move it to separate PR

Comment (Morriz) [reply schemas/gen-missing-crd-schemas.sh:5]: Why? Not conflicting at all. Please don't create work for no reason

File: src/common/envalid.ts [hunk 13]
```diff
@@ -20,32 +30,29 @@ const cliEnvSpec = {
...
-  }
-  path = `${pEnv.ENV_DIR}/.secrets`
-  if (existsSync(path)) {
-    const result = config({ path })
-    if (result.error) console.error(result.error)
-    pEnv = { ...result.parsed, ...pEnv }
-  }
-  return cleanEnv(pEnv, spec)
+  // load local .env and $ENV_DIR/.secrets if we have it
+  ;[`${process.cwd()}/.env`, `${pEnv.ENV_DIR}/.secrets`].forEach((path) => {
```
Comment (j-zimnowoda) [inline src/common/envalid.ts:44]: what does the `;` stand for?

**Comment by j-zimnowoda** [comment]
There is something wrong with sops. Many times installation of otomi has failed due to the following error
```
 Error: panic: runtime error: index out of range [0] with length 0
```
Full log can be found here: https://github.com/redkubes/otomi-core/actions/runs/3437043041/jobs/5731299234

I am wondering if it has to something with encrypting empty files issue? https://github.com/mozilla/sops/issues/483
Interestingly this issue does not occur on the main branch

More 
Today I tried to deploy few times Otomi from

File: src/common/k8s.ts (hunk 5 above)
Comment (Morriz) [reply src/common/k8s.ts:62]: no, those are separately used

Comment (Morriz) [reply src/common/k8s.ts:62]: if you care to rename env vars in another PR, be my guest...

File: src/common/values.ts (hunk 7 above)
Comment (Morriz) [reply src/common/values.ts:123]: to stay consistent between core and api to avoid diff

File: src/common/utils.ts (hunk 6 above)
Comment (Morriz) [reply src/common/utils.ts:27]: to stay consistent between core and api to avoid diff

File: src/common/values.ts (hunk 8 above)
Comment (Morriz) [reply src/common/values.ts:145]: values might not be ready at that stage but the teamConfig is needed?

Comment (Morriz) [reply src/common/values.ts:145]: should be self explanatory

File: .env.sample (hunk 9 above)
Comment (Morriz) [reply .env.sample:4]: done

File: charts/team-ns/templates/argocd/argocd-project.yaml (hunk 11 above)
Comment (Morriz) [reply charts/team-ns/templates/argocd/argocd-project.yaml:58]: bugfix discovered after adding correct input for tests

File: src/common/envalid.ts (hunk 13 above)
Comment (Morriz) [reply src/common/envalid.ts:44]: ask eslint...but I assume it is for correct compiling to have separation between lines

**Comment by Morriz** [comment]
> There is something wrong with sops. Many times installation of otomi has failed due to the following error
> 
> ```
>  Error: panic: runtime error: index out of range [0] with length 0
> ```
> 
> Full log can be found here: https://github.com/redkubes/otomi-core/actions/runs/3437043041/jobs/5731299234
> 
> I am wondering if it has to something with encrypting empty files issue? [mozilla/sops#483](https://github.com/mozilla/sops/issues/483) Interestingly this issue does not occur on the main branch
> 

Those logs show nothing. Please check test files then. I am not up to date with what is going on there. I am pretty sure no empty secret files get generated. Might be another error.

File: src/common/k8s.ts (hunk 5 above)
Comment (Morriz) [reply src/common/k8s.ts:62]: Good idea, but it is a bit of a stint to make envalid deal with logic. Let's keep tha for another time as it's not changed in this PR

**Comment by Morriz** [comment]
@j-zimnowoda my bad, still found empty secret files so removed them.

File: src/common/values.ts (hunk 8 above)
Comment (j-zimnowoda) [reply src/common/values.ts:145]: but it is not. I think adding comment like 
```
# values might not be ready at that stage but the teamConfig is needed
```

Would make the work

File: .values/env/teams/.gitinclude (hunk 10 above)
Comment (j-zimnowoda) [reply .values/env/teams/.gitinclude:1]: If it is a place holder then the convention is to use `.gitkeep`

File: schemas/gen-missing-crd-schemas.sh (hunk 12 above)
Comment (j-zimnowoda) [reply schemas/gen-missing-crd-schemas.sh:5]: but it is delaying feature delivery

**Review by j-zimnowoda — APPROVED** [review]
Remarkable feature. Respect!
